### PR TITLE
 Add Cluster Internal Service Error code along with runcontext

### LINKED ISSUE #5485: internal server error on helm install should be system error
@PriyaModali saw an _Internal server error_ when deploying a Helm chart in Kokoro, but Skaffold mis-attributes the error as a user error (DEPLOY_HELM_USER_ERR):

```
    dev_test.go:265: event received timestamp:<seconds:1614793438 nanos:168027477 > event:<deployEvent:<status:"Failed" err:"install: exit status 1" errCode:DEPLOY_HELM_USER_ERR actionableErr:<errCode:DEPLOY_HELM_USER_ERR message:"install: exit status 1" > > > entry:"Deploy failed"
    dev_test.go:265: event received timestamp:<seconds:1614793438 nanos:168871950 > event:<devLoopEvent:<status:"Failed" err:<errCode:DEPLOY_HELM_USER_ERR message:"install: exit status 1" > > > entry:"Update failed with error code DEPLOY_HELM_USER_ERR"
exiting dev mode because first deploy failed: install: exit status 1
time="2021-03-03T17:43:59Z" level=info msg="Ran [skaffold dev --namespace skaffoldq6hn5 --default-repo gcr.io/k8s-skaffold --port-forward --rpc-port 61893] in 23.358 seconds"
    dev_test.go:251: timed out waiting for port forwarding event
    helper.go:184: Skaffold log:
         Listing files to watch...
        >  - gke-loadbalancer
        > Generating tags...
        >  - gke-loadbalancer -> gcr.io/k8s-skaffold/gke-loadbalancer:v1.20.0-64-gd962ed65f
        > Checking cache...
        >  - gke-loadbalancer: Found Remotely
        > Tags used in deployment:
        >  - gke-loadbalancer -> gcr.io/k8s-skaffold/gke-loadbalancer:v1.20.0-64-gd962ed65f@sha256:e496ba5bd8562eeb30e09a14d24844aaee13598af5bf83ac5c4266efd6f23d90
        > Starting deploy...
        > Helm release skaffold-helm-skaffoldq6hn5 not installed. Installing...
        > Error: an error on the server ("Internal Server Error: \"/api/v1/namespaces/skaffoldq6hn5/services\": the server is currently unable to handle the request") has prevented the request from succeeding (post services)
```

### Information

- Skaffold version: 
- Operating system: 
- Installed via: 
- Contents of skaffold.yaml:

```yaml
<paste your skaffold.yaml here>
```

### Steps to reproduce the behavior

1. a clonable repository with the sample skaffold project
2. `skaffold <command>`
3. ...

### PR DESCRIPTION
**Pull request by tejal29**
Fixes #5485

**Review Order**
1) #5532  - event refactor to use `Config` instead of variables from config
2) #5537 - to use `runContext` and `Config` for error evaluation instead of singleton `skafffoldOpts` package variable.

**Merge After**  #5490, #5532, #5537 

In this PR, summary of changes
1) define  a status code for cluster internal system error
2) change all deployer `userError` to see if the K8 cluster system error.

=== Testing Notes====

Force a Internal System Error via code. 
```
diff --git a/pkg/skaffold/deploy/kubectl/kubectl.go b/pkg/skaffold/deploy/kubectl/kubectl.go
index c9831986d..9fc6a383a 100644
--- a/pkg/skaffold/deploy/kubectl/kubectl.go
+++ b/pkg/skaffold/deploy/kubectl/kubectl.go
@@ -84,6 +84,7 @@ func NewDeployer(cfg Config, labels map[string]string, d *latest.KubectlDeploy)
 // Deploy templates the provided manifests with a simple `find and replace` and
 // runs `kubectl apply` on those manifests
 func (k *Deployer) Deploy(ctx context.Context, out io.Writer, builds []graph.Artifact) ([]string, error) {
+       return nil, userError(fmt.Errorf("Internal Server Error: \"/api/v1/namespaces/skaffoldq6hn5/services\": the server is currently unable to handle the request has prevented the request from succeeding (post services))"))
        var (
                manifests manifest.ManifestList
                err       error

```
and then 
```
LOCAL=true make
```

```
➜  getting-started git:(fix_internal_sys_err) ✗ ../../out/skaffold dev -d gcr.io/tejal-gke1
Listing files to watch...
 - skaffold-example
Generating tags...
 - skaffold-example -> gcr.io/tejal-gke1/skaffold-example:v1.21.0-39-g26f9698fd
Checking cache...
 - skaffold-example: Found Remotely
Starting test...
Tags used in deployment:
 - skaffold-example -> gcr.io/tejal-gke1/skaffold-example:v1.21.0-39-g26f9698fd@sha256:73be4a21d9fc1ac5c32a7c332cb71672a935d9bbd77d4395c0b64cfbc57eb84f
Starting deploy...
Deploy Failed. exiting dev mode because first deploy failed: Internal Server Error: "/api/v1/namespaces/skaffoldq6hn5/services": the server is currently unable to handle the request has prevented the request from succeeding (post services)). Something went wrong with your cluster "gke_tejal-gke1_us-central1-c_dump". Try again.
If this keeps happening please open an issue at https://github.com/GoogleContainerTools/skaffold/issues/new.
```

### CODE REVIEW

File: pkg/skaffold/errors/errors.go [hunk 1]
```diff
@@ -36,16 +37,16 @@ const (
...
 )
 
 var (
-	setOptionsOnce sync.Once
-	skaffoldOpts   config.SkaffoldOptions
+	// Report issue text
+	reportIssueText = fmt.Sprintf("If above error is unexpected, please open an issue %s to report this error", constants.GithubIssueLink)
+
+	setRunContextOnce sync.Once
+	runCtx            runcontext.RunContext
```
Comment (briandealwis) [inline pkg/skaffold/errors/errors.go:47]: Is there any way we can avoid singletons/globals like this?  IMHO if these are needed, they should be pushed into cmd/skaffold/app/cmd and passed in from `ShowAIError()`.

**Review by briandealwis — COMMENTED** [review]
So I think we should avoid passing in a `RunContext` and instead follow the spirit of #4598 with `Config` interfaces.  We can pass in the RunContext as an `interface{}`, but the actual code should be trying to cast it to the appropriate config.

File: pkg/skaffold/errors/errors.go (hunk 1 above)
Comment (tejal29) [reply pkg/skaffold/errors/errors.go:47]: Let me see what i can do here.

File: pkg/skaffold/deploy/error/errors.go [hunk 2]
```diff
@@ -57,3 +73,42 @@ func MissingToolErr(toolName string, err error) string {
...
+	return sErrors.NewProblem(
+		func(err error) string {
+			return fmt.Sprintf("Deploy Failed. %v", err)
+		},
+		proto.StatusCode_DEPLOY_CLUSTER_INTERNAL_SYSTEM_ERR,
+		func(runCtx runcontext.RunContext) []*proto.Suggestion {
+			action := fmt.Sprintf("Something went wrong with your cluster %q", runCtx.KubeContext)
+			if isMinikube(runCtx.KubeContext) {
+				action = fmt.Sprintf("Run minikube status -p %s to check if minikube is running", runCtx.KubeContext)
+			}
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/deploy/error/errors.go:91]: yeah, i think we could add a utility in deploy or error package to show minikube, kind and gke specific AIs

File: cmd/skaffold/app/cmd/cmd.go [hunk 3]
```diff
@@ -268,8 +268,10 @@ func setUpLogs(stdErr io.Writer, level string, timestamp bool) error {
 
 func alwaysSucceedWhenCancelled(ctx context.Context, cfg sErrors.Config, err error) error {
 	// if the context was cancelled act as if all is well
-	if err == nil || ctx.Err() == context.Canceled {
+	if err != nil && ctx.Err() == context.Canceled {
 		return nil
+	} else if err == nil || err == context.Canceled {
+		return err
```
Comment (briandealwis) [inline cmd/skaffold/app/cmd/cmd.go:274]: Is this right?  If `err == context.Canceled` then we'll return `context.Canceled`.

Should this be:
```
  if err == nil || ctx.Err() == context.Canceled || err = context.Canceled {
    return nil
  }
  return sErrors.ShowAIError(cfg, err)
```

**Review by briandealwis — COMMENTED** [review]
I'm missing the "Cluster Internal Service Error code"

File: cmd/skaffold/app/cmd/cmd.go (hunk 3 above)
Comment (tejal29) [reply cmd/skaffold/app/cmd/cmd.go:274]: > I'm missing the "Cluster Internal Service Error code"

This is really strange. 
```
➜  skaffold git:(fix_helm_2) git push d fix_helm_2 
Everything up-to-date
➜  skaffold git:(fix_helm_2) git push d fix_helm_2
```
And i see them locally. let me rebase on #5537 and try again.

Comment (tejal29) [reply cmd/skaffold/app/cmd/cmd.go]: > Is this right? If `err == context.Canceled` then we'll return `context.Canceled`.
> 
> Should this be:
> 
> ```
>   if err == nil || ctx.Err() == context.Canceled || err = context.Canceled {
>     return nil
>   }
>   return sErrors.ShowAIError(cfg, err)
> ```

I am not sure about this condition. On master right now [here](https://github.com/GoogleContainerTools/skaffold/blob/f241521cbcdea86fc93d5f5a9528e048f529d86e/cmd/skaffold/app/cmd/cmd.go#L272) the `alwaysSucceedWhenCancelled` method returns `context.Cancelled` error. 
The test relies on this behavior [here](https://github.com/GoogleContainerTools/skaffold/blob/master/cmd/skaffold/app/cmd/dev_test.go#L109)

I actually went and did you mentioned and suddenly all tests started failing hence i decided to not change it. 
I think, the main, ignore cancelled error below. 
https://github.com/GoogleContainerTools/skaffold/blob/master/cmd/skaffold/skaffold.go#L38

Comment (briandealwis) [reply cmd/skaffold/app/cmd/cmd.go]: That test is checking the result of `doDev()`, which should return `context.Canceled`.

That code in skaffold's main is funny as we should be passing in our context.

My real point here is that `alwaysSucceedWhenCancelled` doesn't always succeed when cancelled :-)

File: proto/enums/enums.proto [hunk 4]
```diff
@@ -324,6 +324,8 @@ enum StatusCode {
     DEPLOY_PARSE_MANIFEST_IMAGES_ERR = 1021;
     // Helm config `createNamespace` not available
     DEPLOY_HELM_CREATE_NS_NOT_AVAILABLE = 1022;
+    // K8 cluster internal system err
```
Comment (briandealwis) [inline proto/enums/enums.proto:327]:
```suggestion
    // Kubernetes cluster reported an internal system error
```

File: cmd/skaffold/app/cmd/cmd.go (hunk 3 above)
Comment (tejal29) [reply cmd/skaffold/app/cmd/cmd.go]: yep added a TODO to fix this. 
#5686

File: pkg/skaffold/deploy/error/errors.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/deploy"
```
Comment (gsquared94) [inline pkg/skaffold/deploy/error/errors.go:23]: it seems a bit odd that the `error` package is taking a dependency on the `deploy` package and not the other way round.

Comment (tejal29) [reply pkg/skaffold/deploy/error/errors.go:23]: Just making sure you did not confuse `skaffold/error` package with  `deploy/error`.


That said, I should be able to revert the dependency. (hopefully no cyclic dependency)

File: pkg/skaffold/deploy/deploy_problems_test.go [hunk 6]
```diff
@@ -65,8 +122,11 @@ func TestSuggestDeployFailedAction(t *testing.T) {
 			if test.isMinikube {
 				cfg.minikube = test.context
 			}
-			actual := suggestDeployFailedAction(cfg)
-			t.CheckDeepEqual(test.expected, actual)
+			actual := sErrors.ShowAIError(&cfg, test.err)
+			fmt.Println(actual.Error(), "\n", test.expected)
```
Comment (tejal29) [inline pkg/skaffold/deploy/deploy_problems_test.go:126]: TODO: Remove this

File: pkg/skaffold/deploy/error/errors.go (hunk 5 above)
Comment (gsquared94) [reply pkg/skaffold/deploy/error/errors.go:23]: I had seen that; it seemed at the moment that `deploy` should import `deploy/error` but now I don't have a strong opinion either way.

File: pkg/skaffold/deploy/deploy_problems.go [hunk 7]
```diff
@@ -26,49 +26,84 @@ import (
 	"github.com/GoogleContainerTools/skaffold/proto/v1"
 )
 
+const (
+	defaultMinikubeProfile = "minikube"
+)
+
+var (
+	ClusterInternalSystemErr = regexp.MustCompile(".*Internal Server Error")
```
Comment (gsquared94) [inline pkg/skaffold/deploy/deploy_problems.go:34]: nit: this is exported only to be used in the `error` package. Maybe just move it there.

Comment (tejal29) [reply pkg/skaffold/deploy/deploy_problems.go]: It was used in `deploy` and `deploy/error` package previously and hence exported.
https://github.com/GoogleContainerTools/skaffold/pull/5491/commits/37603cc58104eac8d1b85963b4c41db8f4e74f9d#diff-a8b4a0e99c70a1b6f90aff684f68c5f25ec7650177441b92ceac2557ae3d9f35R68
and 
https://github.com/GoogleContainerTools/skaffold/pull/5491/commits/37603cc58104eac8d1b85963b4c41db8f4e74f9d#diff-8da1821d1e711320629c6b9807272ea867c0dd81e63186cfb87a9ee087b37465R66

Made private after refactor.

File: pkg/skaffold/deploy/error/errors.go (hunk 5 above)
Comment (tejal29) [reply pkg/skaffold/deploy/error/errors.go]: fixed!

**Review by gsquared94 — APPROVED** [review]

**Pull request merged**